### PR TITLE
Use ServiceLoader for ObjectFactory

### PR DESCRIPTION
**Pull request by nhojpatrick**
## Summary

Is this the type of changes you where expecting for #1450?

## Details

Build fails at spring tests, but before I send more time is this what you where thinking?

## How Has This Been Tested?

Not fully tested yet, tests are failing. Still work in process...

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue).
- [ ] New feature (non-breaking change which adds functionality).
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected).

## Checklist:

- [ ] I've added tests for my code.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: java/src/main/java/io/cucumber/java/ObjectFactoryLoader.java [hunk 1]
```diff
@@ -2,45 +2,55 @@
...
-import io.cucumber.core.reflection.Reflections;
-import io.cucumber.core.reflection.TooManyInstancesException;
 
-import static java.util.Arrays.asList;
+import java.util.Iterator;
+import java.util.ServiceLoader;
 
 public class ObjectFactoryLoader {
+
+    private static final ServiceLoader<ObjectFactory> LOADER = ServiceLoader.load(ObjectFactory.class);
```
Comment (mpkorstanje) [inline java/src/main/java/io/cucumber/java/ObjectFactoryLoader.java]: The service loader is not safe for use by multiple concurrent threads. Runners are executed in parallel and each runner will create it's own backends with object factory. You can either make all fields and methods non-static or move this field into the method.

File: java/src/main/java/io/cucumber/java/ObjectFactoryLoader.java [hunk 2]
```diff
@@ -2,45 +2,55 @@
...
+
     private ObjectFactoryLoader() {
     }
 
     /**
      * Loads an instance of {@link ObjectFactory}. The class name can be explicit, or it can be null.
-     * When it's null, the implementation is searched for in the <pre>cucumber.runtime</pre> packahe.
+     * When it's null, the implementation is searched for in the <pre>cucumber.runtime</pre> package.
      *
+     * @deprecated as of version 4.0.0; use {@code loadObjectFactory() } instead.
```
Comment (mpkorstanje) [inline java/src/main/java/io/cucumber/java/ObjectFactoryLoader.java]: While currently not package private, the ObjectFactoryLoader will be after merging. There is no need to keep a deprecated method around.

**Comment by mpkorstanje** [comment]
> Build fails at spring tests, but before I send more time is this what you where thinking?

Looks exactly as expected! Thanks!

Btw. You'll notice that by adding whitespace changes in files that you otherwise didn't touch you'll now have merge conflicts. While I really appreciate the sentiment of cleaning things up, try to make these changes only in combination with other changes.

**Comment by nhojpatrick** [comment]
yep i auto format style often without thing, sometimes i only do it so it appears in git as a file i need to go back to check later.

undone changes

**Comment by mpkorstanje** [comment]
> Merge remote-tracking branch 'upstream/master' into develop-v5_1450_s…

You may want to revert that commit and instead use upstream/develop-v5

**Comment by mpkorstanje** [comment]
I cherry picked your changes into a03dd01. Cheers!

**Pull request closed without merging**